web/src/reducers/activities.selectors.js

```diff
@@ -0,0 +1,76 @@
+import { createSelector } from 'reselect';
+import { selectApdData } from './apd.selectors';
+
+export const selectActivityKeys = ({ activities: { allKeys } }) => allKeys;
+
+export const selectActivitiesByKey = ({ activities: { byKey } }) => byKey;
+
+export const selectActivityByKey = ({ activities: { byKey } }, { aKey }) =>
+  byKey[aKey];
+
+const selectBudgetForActivity = ({ budget }, { aKey }) =>
+  budget.activities[aKey];
+
+export const makeSelectCostAllocateFFP = () =>
+  createSelector(
+    [selectActivityByKey, selectBudgetForActivity],
```

Comment:
`createSelector` does the magic of creating memoized functions.  The first argument is an array of selectors (it passes state and props into them), and the results are used as parameters into the second argument.  If the parameters are the same between two calls, it just returns the cached result rather than recomputing.

The second argument receives the results of the selectors in the first argument, in the order they're listed.  As far as I can tell, it doesn't have direct access to state or props itself, so it has to get everything from those other selectors.

My rule of thumb on these is if there's no data mutation or computation, use a regular selector (like the first few in this file).  If it ***does*** mutate or compute data, use a memoized selector with `createSelector`.

Reply:
I'd love to capture some of the thinking here in our docs somewhere. It's really smart, but I also think I'd have a hard time understanding what's going on if I were coming into the code cold. (Hell, I kind of have a hard time with it now.)